<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>twostageGWASsurvival</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">twostageGWASsurvival</h1>



<p>The package implements a two stage method for finding significant
interaction effects within a high-dimensional setting where the
covariates outnumber the subjects assuming a Cox proportional hazards
model.</p>
<p>Since the package is not (yet) available on CRAN, installation goes
as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;luc-van-schijndel/twostageGWASsurvival&quot;</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>, <span class="at">build_manual =</span> <span class="cn">TRUE</span>, <span class="at">build_vignettes =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(twostageGWASsurvival)</span></code></pre></div>
<p>The two stage method is, as the name implies, based on two stages.
The first stage tests all covariates individually for a significant
marginal effect. The covariates that are found to be marginally
significant are passed on to the next stage. The second stage then tests
all pairs of marginal significant for a significant interaction effect.
Since we are testing multiple hypotheses, we have to correct for this
fact, which is of course done by applying a multiple hypotheses
correction such as the Bonferroni correction for example. The crucial
part is that the two stages, more specifically the p-values found in the
two stages, are independent of each other. The proof of this is quite
long, but it follows from a thorough analysis of the Fisher information
matrix. By this independence, we only need to apply the multiple
hypotheses correction to the hypotheses tested in the second stage.</p>
<p>By viewing the first stage as a filter of some sort, you get to see
the power of this two stage method. Due to the filter, the number of
hypotheses tested in the second stage decreases quickly as the filter
becomes more strict. The number of pairs tested in the second stage
grows quadratically, so if only one in one hundred covariates are found
to be marginally significant, the second stage tests only one in every
ten thousand pairs.</p>
<p>This allows for a fast computation to a relatively complex problem.
Especially in genetic research, this is a big boon, since the number of
genetic covariates can be as large as several million and consequently,
the number of pairs of covariates can be as large as several trillion.
By applying a strict filter, this number of pairs can be reduced by a
large enough factor that the analysis of these interactions becomes
possible.</p>
<div id="usage" class="section level1">
<h1>Usage</h1>
<p>This filter is specified by the value of the first stage threshold.
This is the threshold the p-values of the first stage need to beat in
order to be passed to the second stage. Under the null hypotheses, these
p-values are uniformly distributed, so we expect that the fraction
specified by the first stage threshold would be marginally significant
and therefore passed to the second stage. This value can be tweaked by
the user to achieve a suitable analysis.</p>
<p>By lowering the first stage threshold, less pairs of covariates will
be tested in the second stage, resulting in a less strict multiple
hypotheses correction and therefore increasing power. This is another
benefit, aside from decreasing computation times. The filter must not be
too strict however, since we do want any actual effects to be detected.
Consider this value therefore properly.</p>
</div>
<div id="example-uses" class="section level1">
<h1>Example uses</h1>
<p>A small example is given by the following datasets:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>survival.dataset <span class="ot">&lt;-</span> survival<span class="sc">::</span><span class="fu">Surv</span>(<span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>),</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>                                   <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>covariate.matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>                             <span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>                             <span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>                             <span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>,</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>                             <span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>                             <span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>                             <span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>,</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>                             <span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">0</span>,</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>                             <span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>,</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>                             <span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>),</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>                           <span class="at">nrow =</span> <span class="dv">10</span>, <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Looking closely at this data, we find that all covariates have a
positive influence on the time to event since we find the highest values
of the covariates for the highest survival times. The main function of
this package, can be used to analyse this tiny dataset assuming a Cox
proportional hazards model:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">twostagecoxph</span>(survival.dataset, covariate.matrix,</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>                        <span class="at">control =</span> <span class="fu">twostagecoxph.control</span>(<span class="at">progress =</span> <span class="dv">0</span>))</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt;                                                                                     </span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt; twostagecoxph(survival.dataset = survival.dataset, covariate.matrix = covariate.matrix, </span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt;     control = twostagecoxph.control(progress = 0))</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; 3 covariates were marginally significant at level 0.05, resulting in 3 second stage tests.</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt; These are the most significant interactions found with their respective p-values:</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt;                        1 x 2   </span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; p-value epistasis      0.585673</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt; amount similar results 0</span></span></code></pre></div>
<p>The object returned from the function has a print method associated
with it, and this method provides a quick summary of the results. The
object itself contains all aspects of the result, such as the vector of
p-values found in the first stage, and the resulting p-values from the
second stage arranged in a sparse matrix.</p>
<p>The function also has parallel processing capabilities, but these
require a parallel back-end to be registered beforehand. Since the
previous example is too small to profit from the multicore capabilities,
let us introduce a toy GWAS dataset that comes with the package:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">str</span>(example_survival_data)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co">#&gt;  &#39;Surv&#39; num [1:200, 1:2] 1.42723+ 0.58487+ 0.88960  0.20406+ 3.85279+ 4.05749+ 1.01883+ 2.83626  1.46397  1.54486+ ...</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt;  - attr(*, &quot;dimnames&quot;)=List of 2</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt;   ..$ : NULL</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt;   ..$ : chr [1:2] &quot;time&quot; &quot;status&quot;</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt;  - attr(*, &quot;type&quot;)= chr &quot;right&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">str</span>(example_snp_data)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co">#&gt;  int [1:200, 1:500] 2 2 2 2 2 2 2 2 2 2 ...</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt;  - attr(*, &quot;dimnames&quot;)=List of 2</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt;   ..$ : chr [1:200] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt;   ..$ : chr [1:500] &quot;SNP_1683421&quot; &quot;SNP_226557&quot; &quot;SNP_1173455&quot; &quot;SNP_1785590&quot; ...</span></span></code></pre></div>
<p>In this synthetic GWAS, we had 200 subjects of which we documented
500 SNPs. These SNPs have a strong correlation structure, as we often
see in covariates derived from the (human) genome. One pair has a build
in influence on the hazard rate according to a Cox proportional hazards
model. Lets register a parallel back-end and use the package to analyse
this dataset:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>doParallel<span class="sc">::</span><span class="fu">registerDoParallel</span>(<span class="dv">2</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">twostagecoxph</span>(example_survival_data, example_snp_data, <span class="at">multicore =</span> <span class="cn">TRUE</span>,</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>                        <span class="at">control =</span> <span class="fu">twostagecoxph.control</span>(<span class="at">progress =</span> <span class="dv">0</span>))</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#&gt;                                                                                     </span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt; twostagecoxph(survival.dataset = example_survival_data, covariate.matrix = example_snp_data, </span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt;     multicore = TRUE, control = twostagecoxph.control(progress = 0))</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt; 191 covariates were marginally significant at level 0.05, resulting in 18145 second stage tests.</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt; These are the most significant interactions found with their respective p-values:</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt;                        SNP_263757 x SNP_222191 SNP_412807 x SNP_120042</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt; p-value epistasis      0.000684222             0.004159757            </span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt; amount similar results 1                       5                      </span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt;                        SNP_412807 x SNP_961485 SNP_348130 x SNP_478023</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt; p-value epistasis      0.007463934             0.052284475            </span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="co">#&gt; amount similar results 0                       0                      </span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="co">#&gt;                        SNP_2224824 x SNP_478023</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="co">#&gt; p-value epistasis      0.065584745             </span></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="co">#&gt; amount similar results 5</span></span></code></pre></div>
<p>The functions finds some significant results one of which was the
interaction used to generate the data. Three pairs of covariates have a
corresponding p-value that is lower than 0.05, so at this level of
significance we find a significant interaction. Moreover, some other
pairs have the same (significant) p-values as other pairs. All of these
are listed in the object:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>result<span class="sc">$</span>most.significant.results<span class="sc">$</span>duplicate.interactions</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co">#&gt; $`SNP_263757 x SNP_222191`</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co">#&gt; SNP_263757 x SNP_2128600 </span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt;             0.0006842215 </span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt; $`SNP_412807 x SNP_120042`</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt;  SNP_412807 x SNP_651758 SNP_412807 x SNP_1487093  SNP_412807 x SNP_840515 </span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt;              0.004159757              0.004159757              0.004159757 </span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#&gt;  SNP_412807 x SNP_737709  SNP_412807 x SNP_892406 </span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt;              0.004159757              0.004159757 </span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co">#&gt; $`SNP_412807 x SNP_961485`</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#&gt; named numeric(0)</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="co">#&gt; $`SNP_348130 x SNP_478023`</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a><span class="co">#&gt; named numeric(0)</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a><span class="co">#&gt; $`SNP_2224824 x SNP_478023`</span></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a><span class="co">#&gt;  SNP_595526 x SNP_478023  SNP_510866 x SNP_478023  SNP_105982 x SNP_478023 </span></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a><span class="co">#&gt;               0.06558475               0.06558475               0.06558475 </span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a><span class="co">#&gt; SNP_2143785 x SNP_478023  SNP_341004 x SNP_478023 </span></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a><span class="co">#&gt;               0.06558475               0.06558475</span></span></code></pre></div>
<p>This means that the interaction of these pairs are so similar to the
interactions displayed in , that the p-values differ too little. These
are therefore reported in a separate part of the object.</p>
</div>
<div id="memory" class="section level1">
<h1>Memory</h1>
<p>This function is written with the intention that it would be used in
a Genome Wide Association Study. In such a study, the amount of
covariates can rise to the millions, which gives rise to problems
concerning memory. To address this, a function has been implemented to
read to covariates in batches from multiple files. This way, the user
has control over how much memory is required for the operation of the
function. The user has to provide the files themselves, and a way for
the function to read them. The large variety of data structures does not
allow for an easy way for the function the adapt the files to something
useable, so the user is therefore required to do some work beforehand.
The files should be structured in such a way, that the user-provided
function results in a matrix where each column corresponds to a
covariate. Additional requirements are described in the documentation.
Lets make some files from the previous example to show how to pass files
to the package:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>number.of.covs <span class="ot">&lt;-</span> <span class="fu">dim</span>(example_snp_data)[<span class="dv">2</span>]</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="co"># Split the covariate matrix into various files.</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>number.of.files <span class="ot">&lt;-</span> <span class="dv">6</span> </span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># 500 is not a multiple of 6, so the last file has less covariates than the other ones</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>temp.snpfile.paths <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="fu">rep</span>(<span class="st">&quot;snpfile&quot;</span>, number.of.files),</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>                               <span class="at">tmpdir =</span> <span class="fu">tempdir</span>(<span class="at">check =</span> <span class="cn">TRUE</span>),</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>                               <span class="at">fileext =</span> <span class="st">&quot;.txt&quot;</span>)</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>indices.matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fu">seq_len</span>(number.of.covs),</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>                           <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">ceiling</span>(number.of.covs<span class="sc">/</span>number.of.files)<span class="sc">*</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>                                 number.of.files <span class="sc">-</span> number.of.covs)),</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>                         <span class="at">ncol =</span> number.of.files)</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a><span class="cf">for</span>(file.num <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>number.of.files){</span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>  indices <span class="ot">&lt;-</span> indices.matrix[,file.num]</span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a>  <span class="fu">write.table</span>(example_snp_data[,indices[<span class="sc">!</span><span class="fu">is.na</span>(indices)]], </span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>              <span class="at">file =</span> temp.snpfile.paths[file.num])</span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>}</span></code></pre></div>
<p>The batch function does allows for the last file to have less
covariates than the other files. This is however the only flexibility
implemented; the function assumes all other files are structured
similarly, e.g. same dimensions, same presence of headers, etc.</p>
<p>The result from the function using batches is then the same as the
one using the basic function:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">print</span>(batched.result <span class="ot">&lt;-</span> </span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>        <span class="fu">batched.twostagecoxph</span>(example_survival_data, temp.snpfile.paths,</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>                              <span class="at">number.of.covariates =</span> number.of.covs,</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>                              <span class="at">read.function =</span> <span class="cf">function</span>(x) <span class="fu">as.matrix</span>(<span class="fu">read.table</span>(x))</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>                              )</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>      )</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co">#&gt;                                                                                     First stage complete. Commencing second stage.                                                                                     Analysis completed. Runtime: 9.89                                                                                    </span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co">#&gt; batched.twostagecoxph(survival.dataset = example_survival_data, </span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co">#&gt;     covariate.filepaths = temp.snpfile.paths, number.of.covariates = number.of.covs, </span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co">#&gt;     read.function = function(x) as.matrix(read.table(x)))</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="co">#&gt; 191 covariates were marginally significant at level 0.05, resulting in 18145 second stage tests.</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="co">#&gt; These are the most significant interactions found with their respective p-values:</span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="co">#&gt;                        SNP_263757 x SNP_222191 SNP_412807 x SNP_120042</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="co">#&gt; p-value epistasis      0.000684222             0.004159757            </span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a><span class="co">#&gt; amount similar results 1                       5                      </span></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="co">#&gt;                        SNP_412807 x SNP_961485 SNP_348130 x SNP_478023</span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a><span class="co">#&gt; p-value epistasis      0.007463934             0.052284475            </span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a><span class="co">#&gt; amount similar results 0                       0                      </span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a><span class="co">#&gt;                        SNP_2224824 x SNP_478023</span></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a><span class="co">#&gt; p-value epistasis      0.065584745             </span></span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a><span class="co">#&gt; amount similar results 5</span></span></code></pre></div>
<p>It is important to pass the necessary function to read the files into
a matrix. The files can be in any format, so long as the
<code>read.function</code> parses the files into the correct matrix,
with the appropriate aforementioned constraints. By default, the
function <code>function(x) as.matrix(read.table(x))</code> is used, so
in this explicit example this parameter was unnecessary, but for clarity
of how to use the function was specified explicitly. Note that we must
provide the paths to the files as a vector, and that we let the function
know how many covariates can be found in all files combined. The
function does allow that to be done automatically, but this requires it
to read through all files individually, which costs time. The user
probably already knows this, so it is best if it is provided. Also note
that this function performs its operations in parallel, just like the
previous example with .</p>
<p>When using this function, a consideration must be made concerning the
amount of covariates in each file. The function always has no more than
the contents of two files in memory on each core, but, it also needs
some additional memory for fitting the model, and saving the results,
etc. It is therefore recommended that no more than one quarter of the
available memory of each core would be used by one file, just to be on
the safe side. Contrary to that, it is also recommended that the files
are as large as possible as every file has some overhead, e.g. creating
the connection. Making the files as large as possible minimizes the
number of files, while not impacting the overall analysis.</p>
</div>
<div id="additional-possibilities" class="section level1">
<h1>Additional possibilities</h1>
<p>This package may be too limited in its implementation for all users.
Since all research is different, we have decided to not try to implement
too much variation, but only provide the most basic functioning package.
If a user needs additional functionality, they are highly encouraged to
implement their own adaptations. In order to aid this, a highly
documented version of the source code is available on the GitHub
repository of this package. This should help you in avoiding the many
pitfalls lurking in implementing this two stage method.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
